### PR TITLE
[Sema] Improving implicit closure capture diagnostic wording

### LINKED ISSUE #58351: [SR-16092] [DiagnosticsQol] Misleading non-sendable capture warning in implicit auto closure
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-16092      |
   |Radar             | None         |
   |Original Reporter | @LucianoPAlmeida      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

main@541f7725e1b0f4ac2cfc1f1b71a90c5a5c813c1d

<https://godbolt.org/z/8abnvqhhr>

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug        |
   |Assignee          | @LucianoPAlmeida      |
   |Priority          | Medium      |

   

   md5: 2d3f027f3cb30df473fa32a2be8104a2

  </details>

**Issue Description:**

Consider the following async let case

``` java
func testAsyncSequence<Seq: AsyncSequence>(_ seq: Seq) async throws where Seq.Element == Int {
    async let result = seq
    // warning: capture of 'seq' with non-sendable type 'Seq' in a `@Sendable` closure
 }
```

The diagnostic is correct because seq is non-sendable and the implementation details is that the rhs of an async let expression is implicitly wrapped in an auto closure.  
But for user perspective this wording may be confusing because there isn't a closure in the source.

The wording should be tailored for implicit non-sendable value captures

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
Improve diagnostic wording for implicit closure captures such as `async let` capture.

Resolves SR-16092.
Fixes #58351

### CODE REVIEW

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -4613,6 +4613,9 @@ ERROR(concurrent_access_of_inout_param,none,
 ERROR(non_sendable_capture,none,
       "capture of %1 with non-sendable type %0 in a `@Sendable` closure",
       (Type, DeclName))
+ERROR(implicit_non_sendable_capture,none,
+      "implicit capture of %1 requires that %0 conforms to `Sendable`",
```
Comment (xedin) [inline include/swift/AST/DiagnosticsSema.def:4617]: I think this explains compiler details but doesn't really help users, how about something like `cannot capture non-sendable type <decl-name> in 'async let' binding <binding-name>' or `'async let' binding <binding-name> cannot capture non-sendable <decl-name>' and a note to add `: Sendable` or something like that.

@hborla WDTY?

Comment (LucianoPAlmeida) [reply include/swift/AST/DiagnosticsSema.def:4617]: Thanks @xedin, I'm making the adjustments. But initially I'm thinking in `capture of <decl-name> with non-sendable type <type> in 'async let' binding` to remain consistent with `non_sendable_capture` one. WDYT?

Comment (xedin) [reply include/swift/AST/DiagnosticsSema.def:4617]: Sounds good which a minor change - if you going to use `type` then print a type and point to the declaration with a note.

**Review by xedin — APPROVED** [review]
Looks good to me but I'd let @kavon have a final word here.

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test

**Comment by LucianoPAlmeida** [comment]
@kavon Friendly ping for a review =]

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test

**Review by kavon — APPROVED** [review]
This looks great. Thank you! Apologies for the late review.

**Comment by kavon** [comment]
@swift-ci please test and merge

**Comment by LucianoPAlmeida** [comment]
> This looks great. Thank you! Apologies for the late review.

No worries @kavon, thanks for the review!

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test Windows Platform

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test MacOS Platform

**Pull request merged**